spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/DistriParameterSynchronizer.scala

```diff
@@ -0,0 +1,366 @@
+/*
+ * Copyright 2016 The BigDL Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.intel.analytics.bigdl.utils
+
+import java.nio.ByteBuffer
+import java.util.concurrent._
+
+import com.intel.analytics.bigdl.parameters.{CompressedTensor, FP16CompressedTensor, SerializerInstance}
+import com.intel.analytics.bigdl.tensor.Tensor
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+import org.apache.commons.lang.exception.ExceptionUtils
+import org.apache.log4j.Logger
+import org.apache.spark.sparkExtension.SparkExtension
+import org.apache.spark.storage.{BlockId, BlockManagerWrapper, StorageLevel}
+
+import scala.collection.mutable
+import scala.reflect.{ClassTag, classTag}
+import scala.collection.JavaConverters._
+
+trait DistriParameterSynchronizer[T] {
+
+  /**
+   * Init synchronization context for new parameter
+   * @param name  identifier for parameter
+   * @param globalSize toal size of parameter
+   * @param priority priority for this parameter
+   */
+  def init(name: String, globalSize: Int, priority: Int = 1): Unit
+
+  /**
+   * put parameter to global
+   * @param name identifier for parameter
+   * @param parameter  parameter to put
+   */
+  def put(name: String, parameter: Tensor[T]): Unit
+
+  /**
+   * get parameter for specific identifier
+   * @param name  identifier for parameter
+   * @return  parameter
+   */
+  def get(name: String): Tensor[T]
+
+  /**
+   * clear the synchronizer
+   */
+  def clear(): Unit
+}
+
+class BlockManagerParameterSynchronizer[T: ClassTag](partitionID: Int,
+                                                     totalPartition: Int)
+                                                    (implicit ev: TensorNumeric[T])
+  extends DistriParameterSynchronizer[T] with Serializable {
+
+  import com.intel.analytics.bigdl.utils.BlockManagerParameterSynchronizer.logger
+
+  private val syncResults: mutable.HashMap[String, FutureTask[Tensor[T]]]
+    = new mutable.HashMap[String, FutureTask[Tensor[T]]]()
+
+  private val taskSize: Int = System.getProperty("bigdl.ParameterSynchronier." +
+    "asyncTaskSize", "100").toInt
+
+  private val clearPoolSize: Int = System.getProperty("bigdl.ParameterSynchronier." +
```

Comment:
why multiple pools?

Reply:
three reasons for this :  
1) There are some long running threads like clear threads and request handling threads
2) Because we do internally and asynchronously, some threads are blocked to wait for other threads result, thus we should do in different pools to avoid blocking
3) Easy to tune the pool size to make it high efficient